ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -618,7 +618,8 @@ static int HandleFileOpOperation(
             }
         }
     }
-    else if (KAUTH_FILEOP_CLOSE == action)
+    else if (KAUTH_FILEOP_CLOSE == action ||
+             KAUTH_FILEOP_OPEN == action)
```

Comment:
Question related to this comment:

>Is it worth changing
if (fileFlaggedInRoot && KAUTH_FILEOP_CLOSE_MODIFIED != closeFlags)
to
if (fileFlaggedInRoot && (KAUTH_FILEOP_CLOSE == action && KAUTH_FILEOP_CLOSE_MODIFIED != closeFlags)
>
>Is it safe to assume we won't have clashes with KAUTH_FILEOP_CLOSE_MODIFIED on OPEN

With the current changes it's not clear if `KAUTH_FILEOP_CLOSE` or `KAUTH_FILEOP_OPEN` is responsible for sending the `MessageType_KtoU_NotifyFileCreated` message.  

Should it to be exclusively the responsibility of `KAUTH_FILEOP_OPEN`?

Reply:
> Should it to be exclusively the responsibility of `KAUTH_FILEOP_OPEN`?

My opinion is that new file notifications should exclusively be the responsibility of `KAUTH_FILEOP_OPEN`.  We know that `KAUTH_FILEOP_CLOSE` is sometimes too late, and restricting the callback to a specific notification will make that notification more predictable and easier to debug (if it's not delivered when expected).

Reply:
For *new file* notifications I agree we should stick with `KAUTH_FILEOP_OPEN` only. I don't think there's any situation where the open wouldn't catch it but the close event would.

For *modified* notifications note that `KAUTH_FILEOP_OPEN` is also delivered for files opened for *reading*, and there is no way for our code to know which kind of access it was opened for. So this will almost certainly generate a ton of false positives. The only way around that would be to correlate `KAUTH_VNODE_WRITE_DATA` events with `KAUTH_FILEOP_OPEN`. If we get a `KAUTH_VNODE_WRITE_DATA` event that is immediately followed by a `KAUTH_FILEOP_OPEN` from the same thread and PID, we can be fairly certain the open was for writing. The question there is if we really want to be keeping track of that kind of state. `KAUTH_FILEOP_CLOSE` of course has the disadvantage of potentially being delivered quite late if the handle is open for long.

Reply:
@wilbaker testing this out now, but I separated the OPEN into it's own flow as discussed.  There is some overlap with CLOSE (both do the directory check and ShouldProcess check), but I think their behavior is very easy to read now.

@pmj thanks for the extra info regarding modifications and this was actually what I was going to look at next.  It looks like if we get` KAUTH_VNODE_WRITE_DATA` we Hydrate a file.  Could we instead use that flag to say a file has been modified?

Reply:
@pmj that makes sense thanks!  As long as we don't expect tons of false positives I would think that's better than missing the event.  I'll test it out and see how severe things look.  Sounds like the other alternative is pairing with `KAUTH_FILE_OPEN` which as you said becomes more involved.

Reply:
> As long as we don't expect tons of false positives I would think that's better than missing the event. I'll test it out and see how severe things look. 

This would actually be more consistent with our behavior on Windows from the perspective of VFS4G.  ProjFS on Windows delivers a PreConvertToFull callback when a file is opened for write (even if no writes actually occur).

> Waiting for a response from user space on KAUTH_VNODE_WRITE_DATA could have a measurable performance impact. As an optimisation we could consider marking this and possibly some other notifications as not requiring a response from the provider, as I don't think we really need to wait in this case. This will require some not entirely trivial changes to both kext and lib.

To provide a bit more context here the scenario we're looking to fix is:

```
<modify placeholder file>
git status
```

The `git status` should report the placeholder file as modified, but today we detect the modification in the `KAUTH_FILEOP_CLOSE` event (which @jeschu1 has proven can be too late).

It's also worth noting that for correctness we only have to deliver this event (PreConvertToFull) for files that are marked as placeholders (and only once per-file).  Reducing the scope of the callback to these specific files will help reduce the performance impact of making blocking calls to the provider.

(Also as an FYI - #209 is tracking this work)

Reply:
Since it's somewhat related to this change I added a functional test OpenFileAndWriteWithoutClose demonstrating the modify issue.  I've set it to Ignore since we currently don't have a fix.  I'm going to resolve this thread since it's an issue we'll tackle in another PR.